CI: make sure all commits in a PR are buildable.

chopikus opened the pull request:
Before the commit, the build workflow only checks the last commit at the time of a trigger. This commit updates the CI, so that every build check is done on all PR commits.

See the issue #23 in finch-flight-software for more details.

chopikus commented:
> > Nice! Did you test this somewhere?
> 
> https://github.com/utat-ss/build-yml-test

Thanks for adding this link! I tested the .yml on 2 different branches, one of the commits is broken in the first PR, the other PR has a force push after correcting a mistake.

mshinjo reviewed (commented):
Thanks for this @chopikus !
Two very minor comments:
1. get-commits is currently defined as a job, which means it counts as one GitHub Actions check. Is it possible not to define it as a job (so that we don't get a green checkmark for it)?
2. Your Git commit message includes a link to this GitHub repository’s issue. Could we avoid mentioning it in the commit message? The reason is that this repository might not always be hosted at utat-ss/finch-flight-software, or moved in the future. I prefer to keep a clear separation between Git metadata and this GitHub repository's metadata. Please do mention issues in PRs, though. 
(This is just my preference, not a strict rule we have defined.)
cc @alexapostolu What do you think?

mshinjo commented on .github/workflows/build.yml:
```diff
@@ -55,8 +95,18 @@ jobs:
-      - name: Build app
+      - name: Builds
         run: |
+          set -e
```
Do we want this? I'm not sure how `set -e` behaves in GitHub Actions. Will it force-quit all queued and running jobs (does it cancel runs of other build jobs, like `pay`, if `obc` fails)?

alexapostolu reviewed (changes requested):
Looks good. But if we have `set -e` and we're checking all the commits, if one commit fails won't that stop the checking for the other commits as well?

alexapostolu commented:
Even if the link to the issue is invalid, the number will still be the same right? So I think it's fine cause it's still tracible back to the original issue

mshinjo commented:
> Even if the link to the issue is invalid, the number will still be the same right? So I think it's fine cause it's still tracible back to the original issue

This GitHub repository itself might get deleted (not likely, but possible).

chopikus commented:
> 1. get-commits is currently defined as a job, which means it counts as one GitHub Actions check. Is it possible not to define it as a job (so that we don't get a green checkmark for it)?

I defined it as a separate job to make the list of commits consistent. Otherwise it would be a step, and we would query the list 5 times (once for each board). In that case there is a small chance it will be different for some boards compared to the others.

chopikus commented:
> Looks good. But if we have `set -e` and we're checking all the commits, if one commit fails won't that stop the checking for the other commits as well?

Yes, that will stop checking the other commits and fail, and it will cancel other boards as well. I think right now the board behaviour is the same.

fail-fast is true by default in Actions:
https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast

mshinjo commented:
> Do we need to run all boards and all commits even if some of the cases fail?

Yes, that would be useful.

Example:
Assumption: OBC and PAY wouldn't pass the checks.
OBC check fails. So, the author fixes OBC without knowing that PAY check does not pass, because PAY check was canceled.
Then, the author realizes that PAY has issues as well.
That's two steps. We want to know what's wrong without triggering CI multiple times.

chopikus commented:
> > In that case there is a small chance it will be different for some boards compared to the others.
> 
> Could you clarify what you mean by this? Why would they be different?

I am thinking of this situation:
* one of the jobs in the matrix start (for example obc)
* another commit is pushed
* then another matrix job starts (pay)

Then pay and obc would have different commit lists

mshinjo commented:
> > > In that case there is a small chance it will be different for some boards compared to the others.
> > 
> > 
> > Could you clarify what you mean by this? Why would they be different?
> 
> Suppose we don't use a separate job to query commits, so we get the commit list each time for every board.
> 
> Imagine this situation (in chronological order):
> 
> * GH starts the `nucleo_g431rb` job
> * the `nucleo_g431rb` job gets the commit list (let's say [hash1])
> * then we push another commit, hash2
> * GH starts another matrix job for `nucleo_h743zi`
> * the `nucleo_h743zi` job gets the commit list, and it will be [hash1, hash2].
> 
> In this case, because we pushed a commit right between the jobs, `nucleo_g431rb` and `nucleo_h743zi` can have different commit lists.
> 
> If we query the commit list beforehand, this situation is not possible. ~sorry that's way too many edits~


> > > In that case there is a small chance it will be different for some boards compared to the others.
> > 
> > 
> > Could you clarify what you mean by this? Why would they be different?
> 
> Suppose we don't use a separate job to query commits, so we get the commit list each time for every board.
> 
> Imagine this situation (in chronological order):
> 
> * GH starts the `nucleo_g431rb` job
> * the `nucleo_g431rb` job gets the commit list (let's say [hash1])
> * then we push another commit, hash2
> * GH starts another matrix job for `nucleo_h743zi`
> * the `nucleo_h743zi` job gets the commit list, and it will be [hash1, hash2].
> 
> In this case, because we pushed a commit right between the jobs, `nucleo_g431rb` and `nucleo_h743zi` can have different commit lists.
> 
> If we query the commit list beforehand, this situation is not possible. ~sorry that's way too many edits~

```
on:
  pull_request:
    types:
      - opened
      - reopened
      - synchronize
```
Doesn't synchronize re-trigger entire workflows? This configuration should cancel all running or queued jobs and re-run the whole workflow."

mshinjo commented:
> > > > In that case there is a small chance it will be different for some boards compared to the others.
> > > 
> > > 
> > > Could you clarify what you mean by this? Why would they be different?
> > 
> > 
> > Suppose we don't use a separate job to query commits, so we get the commit list each time for every board.
> > Imagine this situation (in chronological order):
> > ```
> > * GH starts the `nucleo_g431rb` job
> > 
> > * the `nucleo_g431rb` job gets the commit list (let's say [hash1])
> > 
> > * then we push another commit, hash2
> > 
> > * GH starts another matrix job for `nucleo_h743zi`
> > 
> > * the `nucleo_h743zi` job gets the commit list, and it will be [hash1, hash2].
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In this case, because we pushed a commit right between the jobs, `nucleo_g431rb` and `nucleo_h743zi` can have different commit lists.
> > If we query the commit list beforehand, this situation is not possible. ~sorry that's way too many edits~
> 
> Regarding the commit list job, in simpler terms 😄 If we don't query the commit list in a separate job, we have to do it multiple times concurrently (once for each board in the matrix). The output might not be the same for all boards.

I believe whenever there is a new push, a new set of workflows are triggered and only their results are used to determine whether merging criteria have passed. I don't think we need to get a list of commits manually. Every time there's a push, all jobs are re-run for the new commit, whether older jobs were running, queued, or even finished, and only the latest run counts. In your example, once hash2 is pushed, GitHub Actions starts a completely new run, which includes a new nucleo_g431rb job along with nucleo_h743zi. All jobs in that run will share the same context.

chopikus commented:
> > > > > In that case there is a small chance it will be different for some boards compared to the others.
> > > > 
> > > > 
> > > > Could you clarify what you mean by this? Why would they be different?
> > > 
> > > 
> > > Suppose we don't use a separate job to query commits, so we get the commit list each time for every board.
> > > Imagine this situation (in chronological order):
> > > ```
> > > * GH starts the `nucleo_g431rb` job
> > > 
> > > * the `nucleo_g431rb` job gets the commit list (let's say [hash1])
> > > 
> > > * then we push another commit, hash2
> > > 
> > > * GH starts another matrix job for `nucleo_h743zi`
> > > 
> > > * the `nucleo_h743zi` job gets the commit list, and it will be [hash1, hash2].
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > In this case, because we pushed a commit right between the jobs, `nucleo_g431rb` and `nucleo_h743zi` can have different commit lists.
> > > If we query the commit list beforehand, this situation is not possible. ~sorry that's way too many edits~
> > 
> > 
> > Regarding the commit list job, in simpler terms 😄 If we don't query the commit list in a separate job, we have to do it multiple times concurrently (once for each board in the matrix). The output might not be the same for all boards.
> 
> I believe whenever there is a new push, a new set of workflows are triggered and only their results are used to determine whether merging criteria have passed. I don't think we need to get a list of commits manually. Every time there's a push, all jobs are re-run for the new commit, whether older jobs were running, queued, or even finished, and only the latest run counts. In your example, once hash2 is pushed, GitHub Actions starts a completely new run, which includes a new nucleo_g431rb job along with nucleo_h743zi. All jobs in that run will share the same context.

You are right :slightly_smiling_face: 

I removed the get-commits job and simplified the log messages, they were a bit redundant.

Example run: https://github.com/utat-ss/build-yml-test/actions/runs/15438194267/job/43449476946?pr=5

chopikus commented:
@mshinjo

mshinjo commented on .github/workflows/build.yml:
```diff
@@ -55,8 +62,45 @@ jobs:
+          for commit in $commits; do
+            echo "::group::Building $commit"
+
+            git checkout $commit
+            west -v build -b ${{ matrix.board }} ${{ matrix.path }} -p
+            build_status=$?
```
Can we run `rm -rf build/` right before building?

mshinjo commented on .github/workflows/build.yml:
```diff
@@ -55,8 +62,45 @@ jobs:
+
+          echo "Commit list: $commits"
+
+          failed_commits=""
+          for commit in $commits; do
```
Add a line between these two lines.

mshinjo commented on .github/workflows/build.yml:
```diff
@@ -55,8 +62,45 @@ jobs:
+        env:
+          GH_TOKEN: ${{ github.token }}
         run: |
+          set +e
```
Can we do something like below and not apply `set +e` here?
```
set +e
west -v build -b ${{ matrix.board }} ${{ matrix.path }} -p
build_status=$?
set -e
```

We want to stop the workflow unless it's a build error.

mshinjo commented on .github/workflows/build.yml:
```diff
@@ -10,6 +10,10 @@ on:
     branches:
       - main
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
```
Can we remove this for now? It might be beneficial to allow older workflows to run regardless. If we decide we want this later, we should apply it to all relevant workflows in a separate PR.

chopikus replied:
Yeah, sure! The script also prints the number of commits now to make it more obvious in a rare situation that the commit list became inconsistent in a past run.

chopikus replied:
Sure!

mshinjo commented:
@chopikus I've added four comments; everything else looks good!

mshinjo commented on .github/workflows/build.yml:
```diff
@@ -55,8 +58,49 @@ jobs:
+          for commit in $commits; do
+            echo "::group::Building $commit"
+
+            git checkout $commit
+            set +e
+            rm -rf build/
+            west -v build -b ${{ matrix.board }} ${{ matrix.path }} -p
+            build_status=$?
+            set -e
```
`set +e` right before build, because if rm -rf build/ fails, we don't want to run the build.

```
git checkout $commit
rm -rf build/
set +e
west -v build -b ${{ matrix.board }} ${{ matrix.path }} -p
build_status=$?
set -e
```

mshinjo replied:
`rm -rf` returns 0 even if the directory doesn’t exist

chopikus replied:
Yeah I checked, it returns 0 when the dir doesn't exist.

```
rm --help
Usage: rm [OPTION]... [FILE]...
Remove (unlink) the FILE(s).

  -f, --force           ignore nonexistent files and arguments, never prompt
```
